kafka-client/src/main/java/dev/responsive/kafka/internal/stores/DuplicateKeyBuffer.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.stores;
+
+import dev.responsive.kafka.internal.db.KeySpec;
+import dev.responsive.kafka.internal.utils.Iterators;
+import dev.responsive.kafka.internal.utils.Result;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.NavigableMap;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.stream.Collectors;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.state.KeyValueIterator;
+
+public class DuplicateKeyBuffer<K extends Comparable<K>> implements SizeTrackingBuffer<K> {
+
+  private final NavigableMap<K, LinkedList<Result<K>>> buffer;
+  private final KeySpec<K> extractor;
+  private long bytes = 0L;
+  private int size = 0;
+
+  public DuplicateKeyBuffer(final KeySpec<K> extractor) {
+    this.extractor = Objects.requireNonNull(extractor);
+    this.buffer = new ConcurrentSkipListMap<>();
+  }
+
+  @Override
+  public long sizeInBytes() {
+    return bytes;
+  }
+
+  @Override
+  public int sizeInRecords() {
+    return size;
+  }
+
+  @Override
+  public void put(final K key, final Result<K> value) {
+    // ignore tombstones since deletes are a no-op for duplicate stores
+    if (!value.isTombstone) {
+      bytes += value.size(extractor);
+      ++size;
+
+      buffer.computeIfAbsent(key, k -> new LinkedList<>()).add(value);
```

Comment:
Typically unless there are going to be insertions/deletions from the middle of the list `ArrayList` is almost always a better choice. The memory overhead per-element in a LinkedList is so significant that the memory savings of unused elements in the array list are shadowed (not to mention performance of iterating a LinkedList is significantly worse since there's no necessary memory cohesion)

Reply:
good points, thanks for catching this